internal/buildengine/languageplugin/external_plugin_client.go

```diff
@@ -0,0 +1,178 @@
+package languageplugin
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"syscall"
+
+	"connectrpc.com/connect"
+	"github.com/alecthomas/types/either"
+	"github.com/jpillora/backoff"
+
+	langpb "github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1/language"
+	langconnect "github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1/language/languagepbconnect"
+	"github.com/TBD54566975/ftl/internal/exec"
+	"github.com/TBD54566975/ftl/internal/log"
+	"github.com/TBD54566975/ftl/internal/rpc"
+)
+
+type streamCancelFunc func()
+
+type externalPluginClient interface {
+	getCreateModuleFlags(ctx context.Context, req *connect.Request[langpb.GetCreateModuleFlagsRequest]) (*connect.Response[langpb.GetCreateModuleFlagsResponse], error)
+	createModule(ctx context.Context, req *connect.Request[langpb.CreateModuleRequest]) (*connect.Response[langpb.CreateModuleResponse], error)
+	moduleConfigDefaults(ctx context.Context, req *connect.Request[langpb.ModuleConfigDefaultsRequest]) (*connect.Response[langpb.ModuleConfigDefaultsResponse], error)
+	getDependencies(ctx context.Context, req *connect.Request[langpb.DependenciesRequest]) (*connect.Response[langpb.DependenciesResponse], error)
+
+	build(ctx context.Context, req *connect.Request[langpb.BuildRequest]) (chan either.Either[*langpb.BuildEvent, error], streamCancelFunc, error)
+	buildContextUpdated(ctx context.Context, req *connect.Request[langpb.BuildContextUpdatedRequest]) (*connect.Response[langpb.BuildContextUpdatedResponse], error)
+
+	kill() error
+}
+
+var _ externalPluginClient = &externalPluginImpl{}
+
+type externalPluginImpl struct {
+	cmd    *exec.Cmd
+	client langconnect.LanguageServiceClient
+}
+
+func newExternalPluginImpl(ctx context.Context, bind *url.URL, language string) (*externalPluginImpl, error) {
+	impl := &externalPluginImpl{
+		client: rpc.Dial(langconnect.NewLanguageServiceClient, bind.String(), log.Error),
+	}
+	err := impl.start(ctx, bind, language)
+	if err != nil {
+		return nil, fmt.Errorf("failed to start plugin: %w", err)
+	}
+	return impl, nil
+}
+
+// Start launches the plugin and blocks until the plugin is ready.
+func (p *externalPluginImpl) start(ctx context.Context, bind *url.URL, language string) error {
+	cmdName := "ftl-language-" + language
+	p.cmd = exec.Command(ctx, log.Debug, ".", cmdName, "--bind", bind.String())
+	_, err := exec.LookPath(cmdName)
+	if err != nil {
+		return fmt.Errorf("failed to find plugin for %s: %w", language, err)
+	}
+
+	runCtx, cancel := context.WithCancel(ctx)
+
+	cmdErr := make(chan error)
```

Comment:
Wouldn't an errgroup be preferable here?

Reply:
I don't think it'd work, but I might be missing how to do it.
The ping channel closes when ping succeeds or fails. The cmd chan closes if the plugin exits. In normal cases the plugin won't exit until `plugin.Kill()` is called, but can happen early for whatever reason.
So the `select` is looking for whichever of those 2 happens first, not waiting for both channels.